### PR TITLE
Deprecated as warning

### LINKED ISSUE #40604: List @deprecated strikethroughs in Problems tab
The new deprecated strikethroughs in TS code are great. But to make it easier to find and address a series of them, it would be nice to have them also appear in the Problems tab (as an optional view, controlled by a settings flag).

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #40604 

IMO. That should not an error but an warning because it's just suggestions.

image

### CODE REVIEW

**Comment by morganwdavis** [comment]
Perfect. Thank you!

**Comment by amcasey** [comment]
@orta VS has no problem with warnings.  I think the bigger issue is that we've repeatedly discussed giving some diagnostics warning severity and decided against it.  I cannot, unfortunately, recall our most recent line of reasoning.  @DanielRosenwasser?

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -4729,6 +4729,10 @@
         "category": "Message",
         "code": 6800
     },
+    "Issue an warning if declaration is deprecated": {
```
Comment (juanjoms) [inline src/compiler/diagnosticMessages.json:4732]: Typo: should be "Issue **a** warning if the declaration is deprecated"

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1065,16 +1065,23 @@ namespace ts {
...
-            addErrorOrSuggestion(isError, "message" in message ? createDiagnosticForNode(location, message, arg0, arg1, arg2, arg3) : createDiagnosticForNodeFromMessageChain(location, message)); // eslint-disable-line no-in-operator
+            const category = isError ? DiagnosticCategory.Error : DiagnosticCategory.Suggestion;
+            addErrorOrWarningOrSuggestion(category, "message" in message ? createDiagnosticForNode(location, message, arg0, arg1, arg2, arg3) : createDiagnosticForNodeFromMessageChain(location, message)); // eslint-disable-line no-in-operator
+        }
+
+        function wraningOrSuggestion(isWraning: boolean, location: Node, message: DiagnosticMessage | DiagnosticMessageChain, arg0?: string | number, arg1?: string | number, arg2?: string | number, arg3?: string | number): void {
+            const category = isWraning ? DiagnosticCategory.Warning : DiagnosticCategory.Suggestion;
+            addErrorOrWarningOrSuggestion(category, "message" in message ? createDiagnosticForNode(location, message, arg0, arg1, arg2, arg3) : createDiagnosticForNodeFromMessageChain(location, message)); // eslint-disable-line no-in-operator
```
Comment (sandersn) [inline src/compiler/checker.ts:1084]:
```suggestion
        function warningOrSuggestion(isWarning: boolean, location: Node, message: DiagnosticMessage | DiagnosticMessageChain, arg0?: string | number, arg1?: string | number, arg2?: string | number, arg3?: string | number): void {
            const category = isWarning ? DiagnosticCategory.Warning : DiagnosticCategory.Suggestion;
            addErrorOrWarningOrSuggestion(category, "message" in message ? createDiagnosticForNode(location, message, arg0, arg1, arg2, arg3) : createDiagnosticForNodeFromMessageChain(location, message)); // eslint-disable-line no-in-operator
```

**Review by sandersn — CHANGES_REQUESTED** [review]
I don't like this change as-is since it adds a flag, but maybe it's time we had the warnings/errors/suggestions discussion again.

(Perhaps) notably, this particular bug is a non-issue in emacs since it shows suggestions in the error list. VS may be similar?

**Comment by sandersn** [comment]
image

**Comment by sheetalkamat** [comment]
This should definitely follow unused locals scenario and be error instead of warning as we have stayed away from warning so far and matches what we do for unused locals

**Comment by morganwdavis** [comment]
Adding one last comment...  As an everyday developer, I really don't care if these deprecations are considered errors or warnings.  As long as they get listed in the Problems tab in VSCode, so we don't have to visually scan over every source file looking through strikethroughs or perform manual searches on deprecations we know to search on, that would be ideal. If controversary over how those are presented can be avoided by having a multi-state VSCode setting (List deprecations in Problems tab: Ignore, Error, Warning), then great -- everyone's happy. The main goal should be to provide a convenient list of alerts we can identify in a project and attack efficiently. Thanks.

**Comment by morganwdavis** [comment]
Maybe flagging deprecations in VSCode as "Infos" is the solution? Notice how the Code Spelling Checker extension does it:

image

This gets the job done nicely and isolates these from warnings and errors.

**Comment by Kingwl** [comment]
@sheetalkamat 

So should we use error insted of warning?

**Comment by sheetalkamat** [comment]
I think error under noDeprecated seems like similar to locals but @DanielRosenwasser should comment on that.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -1084,12 +1084,13 @@ namespace ts {
 
         function addErrorOrSuggestion(isError: boolean, diagnostic: DiagnosticWithLocation) {
             if (isError) {
-                diagnostics.add(diagnostic);
+                diagnostics.add({ ...diagnostic, category: DiagnosticCategory.Error });
```
Comment (sheetalkamat) [inline src/compiler/checker.ts:1087]: Instead of this you need to mark the actual error as error by default

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

**Review by sandersn — CHANGES_REQUESTED** [review]
The code looks good now, but I still have two issues:

1. Does making this an error get rid of the strikethrough annotations in Code?
2. If non-Code editors (emacs and VS, at least) can show suggestions in a list, then it seems like the correct fix is there, rather than in Typescript. Especially since listing suggestions would be a generally useful feature -- not just for deprecated things.

**Comment by Kingwl** [comment]
For 1. errorOrSuggestions has rerwrite the type as suggestion. And in the tests it's also suggestion. So i think it's ok.

**Comment by Kingwl** [comment]
For 2. Is that means It's should be handle in code too？ CC: @mjbvz

File: tests/baselines/reference/tsConfig/Default initialized TSConfig/tsconfig.json [hunk 4]
```diff
@@ -40,6 +40,7 @@
     // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
     // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */
     // "noUncheckedIndexedAccess": true,      /* Include 'undefined' in index signature results */
+    // "noDeprecated": true,                  /* Issue a error if declaration is deprecated */
```
Comment (amcasey) [inline tests/baselines/reference/tsConfig/Default initialized TSConfig/tsconfig.json:43]: For consistency with the other comments, consider something like "Report error when referencing deprecated declaration".

**Comment by amcasey** [comment]
I can't really tell how much demand there is for an error when referring to a deprecated symbol.  Is there an upvoted issue or something?

**Review by sheetalkamat — APPROVED** [review]
This looks good to me apart from @amcasey suggestion about the text of the option

File: src/compiler/checker.ts (hunk 2 above)
Comment (mjbvz) [reply src/compiler/checker.ts]: @sandersn VS Code does not include suggestion level diagnostics in the problems view, only errors, warnings, and information messages

**Comment by mjbvz** [comment]
I moved #40604 to TypeScript to track having tsc generate warnings for deprecated symbol usage. On the VS Code side, we can fairly easily change the level of deprecated diagnostics, taking them from `suggestion` to `warning`, without this PR. However I like the PR's proposed `noDeprecated` flag which I presume would control both editor errors and tsc errors

**Comment by Kingwl** [comment]
Up

**Comment by sandersn** [comment]
After thinking about this again, I dislike it more. I'd much rather see editors handle suggestions as suggestions, and perhaps have `tsc` add a flag to emit suggestions alonside errors.

**Comment by Kingwl** [comment]
Okay, Never mind.

**Comment by Kingwl** [comment]
I want to reopen this thread because checking this outside of tsc is tooooooooooooooo slow.

Some Real-World feedback:
image

**Comment by DanielRosenwasser** [comment]
I'd suggest opening as a new issue, maybe with some context on what that table is demonstrating.

**Comment by sandersn** [comment]
I'd recommend filing the bug on vscode since I think it should be implemented there.

**Comment by Kingwl** [comment]
>  on vscode

Sadly It's not a editor only issue If we want to block the compile progress in some environment, such as in CI.

**Comment by johnwc** [comment]
Is there still no way to tell the compiler to throw an error if it finds code using a deprecated method?

**Comment by HolgerJeromin** [comment]
@johnwc 
There is https://github.com/gund/eslint-plugin-deprecation which is working in my setup.

**Pull request closed without merging**